fix(install): if glibc >= 2.18, use the dynamically linked binary

michel-slm opened the pull request:
#### Description
`install.sh` previously uses the statically linked `unknown-linux-musl` build
for every Linux installation. If the host's `libc` is a recent enough
GNU libc, however,  it should use this rather than the statically linked build.

Signed-off-by: Michel Alexandre Salim <michel@michel-slm.name>

#### Motivation and Context
The dynamic build is smaller, and in case there are security vulnerabilities affecting the `libc` implementation, users who don't upgrade to the latest static build will get the fix for free when their distribution releases a security update.

Closes #5361

#### How Has This Been Tested?
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

Tested by:
- short-circuiting `install.sh` to print the result of `detect_platform` and exiting
- testing on Vagrant on the following distributions and verifying that the `unknown-linux-gnu` build works:
  - Arch Linux
  - CentOS 8
  - CentOS Stream 8
  - Debian testing
  - Fedora
  - Ubuntu 20.04

#### Checklist:
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

I don't think there are tests that cover this yet. I've updated the inline docs in `install.sh`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starship/starship/2612)
<!-- Reviewable:end -->

michel-slm committed:
fix: [install.sh] if glibc >= 2.18, use the dynamically linked binary

`install.sh` previously uses the statically linked `unknown-linux-musl` build
for every Linux installation. If the host's `libc` is a recent enough
GNU libc, however,  it should use this rather than the statically linked build.

Includes documentation from the PR conversation about how to recompute the
minimum version as needed.

Signed-off-by: Michel Alexandre Salim <michel@michel-slm.name>

vladimyr commented on install/install.sh:
```diff
@@ -193,8 +193,20 @@ detect_platform() {
+      platform="unknown-linux-musl"
+      if [ -f /etc/os-release ]; then
+	# on known-good distributions, use the dynamically compiled bins
+	distro=$(cat /etc/os-release | grep '^ID=' | sed -e 's/^ID=//')
```
Just source it and work directly with `$DISTRIB_ID`?

vladimyr replied:
I pointed you towards neofetch for a good reason :upside_down_face: Here is how they do it:
```bash
            elif [[ -f /etc/os-release || \
                    -f /usr/lib/os-release || \
                    -f /etc/openwrt_release || \
                    -f /etc/lsb-release ]]; then

                # Source the os-release file
                for file in /etc/lsb-release /usr/lib/os-release \
                            /etc/os-release  /etc/openwrt_release; do
                    source "$file" && break
                done

                # Format the distro name.
                case $distro_shorthand in
                    on)   distro="${NAME:-${DISTRIB_ID}} ${VERSION_ID:-${DISTRIB_RELEASE}}" ;;
                    tiny) distro="${NAME:-${DISTRIB_ID:-${TAILS_PRODUCT_NAME}}}" ;;
                    off)  distro="${PRETTY_NAME:-${DISTRIB_DESCRIPTION}} ${UBUNTU_CODENAME}" ;;
                esac
```
Also, you're gonna need lowercase it because here is what I get on Linux Mint 19.2:
```sh
# from /etc/lsb-release
DISTRIB_ID=LinuxMint

# from /etc/os-release
ID=linuxmint
```

vladimyr replied:
And you can extend that (allow)list with Linux Mint because it is fairly popular Ubuntu derivate :wink:

davidkna replied:
I think I prefer the old check https://github.com/starship/starship/commit/2e14d1af5a3b90aedab39180e0851e8f945b9f55

vladimyr replied:
```sh
if [ "$platform" = "linux" ] && ldd /bin/sh | grep -iq musl; then
  platform=unknown-linux-musl
fi
```

michel-slm replied:
so, something like the old check, but flipping it to detect glibc, would be ideal if we also add a check for the glibc version that works in a cross platform way, right?

michel-slm replied:
this seems to work. I'll rework the PR to use it
```
$ ldd --version | head -n1 | sed -e 's/ldd (GNU libc) //'
2.28
```

(first detecting if the string contains GNU libc, of course)

vladimyr replied:
You can simplify it by replacing `head` plus `sed` combo with awk:
```sh
ldd --version | awk 'NR==1 { print $NF }'
# or invoking /usr/lib/libc.so.6 directly
/usr/lib/libc.so.6 | awk 'NR==1 { print $NF }'
```

vladimyr replied:
Well, in that case you can use following checks:
```sh
ldd --version | awk '/GNU libc/ { print $NF }'
# or invoking /usr/lib/libc.so.6 directly
/usr/lib/libc.so.6 | awk '/GNU libc/ { gsub(/\.$/,""); print $NF }'
```

vladimyr replied:
> I can use `awk` if that's preferable

The sole reason why I prefer `awk` here is to minimize calls to external binaries.

davidkna commented:
I think the main issue wasn't whether a distro uses glibc, but whether the system glibc version was compatible.

vladimyr commented:
@davidkna Do you happen to know what's the minimal supported glibc version?

davidkna commented:
It's more about incompatible ABI versions with any other version. Cross uses v2.15, but it might also use whatever ubuntu is at from the CI env.

vladimyr commented:
Reading through https://github.com/rust-lang/libc/issues/1412 `2.28` seems like a safe bet.
Searching for `glibc >= 2.28` on DistroWatch yields a lot of distros: https://distrowatch.com/search.php?pkg=glibc&relation=greaterequal&pkgver=2.28&distrorange=InLatest#pkgsearch

vladimyr commented:
I'm not sure how to interpret this but I'm pasting it here for the convenience:
```
~ 
❯ strings $(which starship) | grep -i glibc
GLIBC_2.2.5
GLIBC_2.3.3
GLIBC_2.3.2
GLIBC_2.15
GLIBC_2.33
GLIBC_2.17
GLIBC_2.14
GLIBC_2.9
GLIBC_2.18
GLIBC_2.3.4
GLIBC_2.3
GLIBC_2.32
```

vladimyr commented:
> @vladimyr yeah, so now the problem is... older versions of a given distro might have a glibc that is too old. Especially for long-lived distros like Debian oldstable or older CentOS releases

So current Debian oldstable is Debian 9 (stretch) and it ships with glibc 2.24 which might be a problem but the current stable Debian 10 (buster) ships with 2.28 which should work. Debian 9 isn't EOLed AFAIK but there is a newer stable version and supporting older versions shouldn't be Starship's priority anyway.

CentOS 7 ships with glibc 2.17 and the latest CentOS 8 ships with 2.28 but there is a whole CentOS situation with that abrupt EOL notice so I guess we shouldn't be concerned too much about it.

I understand your point and IMHO we should support only popular and regularly updated (non-rolling) distros by supporting their latest stable versions.

vladimyr commented:
According to Amos's checker https://gist.github.com/fasterthanlime/17e002a8f5e0f189861c with `MAX_VER` set to `2.28` latest Starship (v0.51) depends on:
```
~/scratchpad 
❯ ./glibc-check.sh ./starship                                                      
Found version 2.14
Found version 2.15
Found version 2.17
Found version 2.18
Found version 2.2.5
Found version 2.3
Found version 2.3.2
Found version 2.3.3
Found version 2.3.4
Found version 2.4
Found version 2.7
Found version 2.9
```

I guess this means the minimum supported glibc version is 2.18?

vladimyr commented:
:warning: Make sure to check it using `shellcheck`:
```
Line 202:
      if [[ "${ldd_verstr}" == "${ldd_glibc_prefix}"* ]]; then
         ^-- SC3010: In POSIX sh, [[ ]] is undefined.
 
Line 205:
        echo $glibc_version
             ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean: (apply this, apply all SC2086)
        echo "$glibc_version"
 
Line 207:
        if [[ "${glibc_version}" > "${min_version}" ]] \
           ^-- SC3010: In POSIX sh, [[ ]] is undefined.
 
Line 208:
           || [[ "${glibc_version}" == "${min_version}" ]]; then
              ^-- SC3010: In POSIX sh, [[ ]] is undefined.
```

vladimyr commented:
> It's more about incompatible ABI versions with any other version. Cross uses v2.15, but it might also use whatever ubuntu is at from the CI env.

Apparently, it used glibc `v2.18` and our [build runner](https://github.com/starship/starship/runs/2177112874) for `v0.51` release was powered by Ubuntu 20.04.2 which ships with glibc `v2.31`.

michel-slm commented:
ah, didn't notice we're using /bin/sh. Will recheck and drop the requirement to 2.18

michel-slm commented:
odd, Fedora's shellcheck 0.7.1 does not detect SC3010 issues when I pass `shellcheck -o all` or `shellcheck -o SC3010`, but pasting in `shellcheck.net` does bring up that issue as well as some pre-existing issues.

michel-slm commented:
@vladimyr fixed the `SC3010` warnings, dropped the min version to 2.18, and added instructions for computing the minimum version based on your comments. Thanks!

vladimyr commented on install/install.sh:
```diff
@@ -193,8 +193,28 @@ detect_platform() {
+      # to avoid linking issues.
+      platform="unknown-linux-musl"
+      ldd_verstr="$(ldd --version | head -n1)"
+      ldd_glibc_prefix="ldd (GNU libc) "
```
The wording varies between vendors:
```sh
# linuxmint
$ ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1.4) 2.27
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

# ubuntu
ldd (Ubuntu EGLIBC 2.19-0ubuntu6.15) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

# arch
ldd (GNU libc) 2.33
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

michel-slm replied:
oh dear. not good. might be worth figuring out how to detect the musl version string instead.

vladimyr commented on install/install.sh:
```diff
@@ -193,8 +193,28 @@ detect_platform() {
+      platform="unknown-linux-musl"
+      ldd_verstr="$(ldd --version | head -n1)"
+      ldd_glibc_prefix="ldd (GNU libc) "
+      if test "${ldd_verstr#*$ldd_glibc_prefix}" != "$ldd_verstr"; then
```
Why `test` instead of `[]`?

michel-slm replied:
no particular reason, I can switch it back

vladimyr commented on install/install.sh:
```diff
@@ -193,8 +193,28 @@ detect_platform() {
+        # then use https://gist.github.com/fasterthanlime/17e002a8f5e0f189861c with
+        # `MAX_VER` set to the version from earlier e.g. GLIBC_2.33 => 2.33
+        min_version="2.18"
+        if [ "${glibc_version}" > "${min_version}" ] \
```
This is redirection unless escaped: `\>` :warning: 
Even then it is lexical comparison between strings rather than numerical one:
```sh
~ $ [ 3.22 \> 3.3 ] && echo 'true' || echo 'false'
false
```

michel-slm replied:
We can't do numerical comparison, sadly, since glibc version numbers might sometimes be of x.y.z form

I guess the cleanest way is to split based on dot and explicitly compare the parts as integers

vladimyr replied:
> I guess the cleanest way is to split based on dot and explicitly compare the parts as integers

Right, plus making sure that missing semver segments default to zero. :+1:

michel-slm replied:
fixed. Right now we can just compare major and minor, if we ever need to compare patch versions we can add that comparison then.

vladimyr commented on install/install.sh:
```diff
@@ -313,6 +333,8 @@ is_build_available() {
 # defaults
 if [ -z "${PLATFORM-}" ]; then
   PLATFORM="$(detect_platform)"
+  echo ${PLATFORM}
+  # exit 0
```
Debugging leftovers I guess? :upside_down_face:

michel-slm replied:
Woops

michel-slm replied:
fixed

michel-slm commented:
@vladimyr 
- used your awk code snippet to extract the version
- since there are so many different glibc version strings, I checked whether `musl` is missing from the first line instead. Also noticed that musl's ldd outputs the version to stderr, not stdout (ugh!) so I redirected stderr to stdout before testing
